### DISCUSSION TITLE
Safety for a production env

### DISCUSSION THREAD

**Discussion by rodolphoasb** [devlikeapro/waha#115] [Q&A] [answered]
Hey, first I would like to say thank you for all your work on this. The lib looks amazing.

I was wondering what are the core technologies under the hood, it seems like we are not using puppeteer right?

I know it's impossible to garantee full stability with a non oficial API, but I would like your opinion on how much you consider this lib a good fit for a Saas where 50+ clients connect their WhatsApp.

**Comment by allburov** [comment, marked as answer]
Hi, thank you for your kind words!

>seems like we are not using puppeteer right?

It depends - there're few engines and 2 of them use puppeter under the hood, so it looks like a usual WhatsApp Web instance from Meta perspective.
**NOWEB** engine uses direct connection via websockets, so you should be more accurate when dealing with that, but it's not hard, actually.
https://waha.devlike.pro/docs/how-to/engines/

>would like your opinion on how much you consider this lib a good fit for a Saas where 50+ clients connect their WhatsApp.

It's pretty real with a proper infrastructure preparation.
I'd suggest few things here:
1. Run about 5 sessions per container if you're using **WEBJS** engine and about 5-10 for *"NOWEB**. It'll keep most of clients stable if something goes wrong with a container. So in you case you'll have to manage 5-10 containers for 50 sessions.
2. Using proxies for sessions - we're going to release simple proxy management in July (so you can define a single proxy or proxies list when container starts) and we'll improve it later so you can define proxy during when you start a session via API (it's not possible right now)
3. Use the right pattern for communicating with clients https://waha.devlike.pro/docs/overview/how-to-avoid-blocking/
4. (optional) Get a PRO subscription so I can consult you in Telegram :D 

I hope it answers your questions, let me know if there's anything I can help with!
